src/Datadog.Trace/Propagation/B3/B3SpanContextPropagator.cs

```diff
@@ -0,0 +1,209 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using Datadog.Trace.Headers;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Propagation.B3
+{
+    /// <summary>
+    /// Class that hanbles B3 style context propagation.
+    /// </summary>
+    internal class B3SpanContextPropagator : Propagator
+    {
+        private const NumberStyles NumberStyle = NumberStyles.HexNumber;
+        private static readonly string UserKeep = ((int)SamplingPriority.UserKeep).ToString(CultureInfo.InvariantCulture);
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<B3SpanContextPropagator>();
+
+        private B3SpanContextPropagator()
+        {
+        }
+
+        /// <summary>
+        /// Gets the singleton instance of the propagator.
+        /// </summary>
+        public static B3SpanContextPropagator Instance { get; } = new B3SpanContextPropagator();
+
+        /// <summary>
+        /// Propagates the specified context by adding new headers to a <see cref="IHeadersCollection"/>.
+        /// This locks the sampling priority for <paramref name="context"/>.
+        /// </summary>
+        /// <param name="context">A <see cref="SpanContext"/> value that will be propagated into <paramref name="headers"/>.</param>
+        /// <param name="headers">A <see cref="IHeadersCollection"/> to add new headers to.</param>
+        public override void Inject(SpanContext context, IHeadersCollection headers)
```

Comment:
should this not be some extension method that reuses `public override void Inject<T>(SpanContext context, T carrier, Action<T, string, string> setter)` ?

Reply:
Excellent idea for refactoring. I refactored propagator to be carrier agnostic - to continue with the pattern added previously, that carrier could be any kind of type that could hold metadata.